cfgov/scripts/add_final_enforcement_data.py

```diff
@@ -0,0 +1,250 @@
+                data[url]['at_risk_groups'] = at_risk_groups
+            except KeyError:
+                logger.info(f"No data in csv data file for {url}\n")
+
+
+def add_statutes(data_file):
+    with open(data_file, 'r', encoding='utf-8-sig') as csv_file:
+        split = csv.reader(csv_file, delimiter=',')
+        # header
+        next(split)
+
+        provisions = ['Deceptive', 'Unfair', 'Abusive']
+
+        o = {}
+
+        for fields in split:
+            (
+                statute,
+                provision,
+                url
+            ) = strip_fields(fields, 1, 2, 3)
+
+            if statute == 'CFPA' and provision in provisions:
+                statute = 'CFPA ' + provision
+
+            try:
+                o[url].append(statute)
+            except KeyError:
+                o[url] = [statute]
+
+        for url in o:
+            data[url]['statutes'] = [
+                {'statute': x} for x in list(dict.fromkeys(o[url]))
+            ]
+
+
+def run(*args):
+    if not args or len(args) != 3:
+        logger.error("error. Use --script-args [RELIEF_PATH] " +
+                     "[PRODUCT_PATH] [STATUTE_PATH] " +
+                     "to specify the location of the data csvs.")
+        sys.exit()
+    add_relief(args[0])
+    add_products(args[1])
+    add_statutes(args[2])
+
+    count = 0
+
+    for page in EnforcementActionPage.objects.all():
+        if not page.live or page.get_parent().title != 'Enforcement Actions':
+            logger.info(f"Skipping unpublished page: {page.slug}\n")
+            continue
+        url = 'https://www.consumerfinance.gov' + page.get_url()
+
+        try:
+            page.public_enforcement_action = data[url][
+                'public_enforcement_action'
+            ]
+            page.initial_filing_date = data[url]['initial_filing_date']
+            page.court = data[url]['court']
+            page.settled_or_contested_at_filing = data[url][
+                'settled_or_contested_at_filing'
+            ]
+
+            page.enforcement_dispositions.set([
+                EnforcementActionDisposition(**x)
+                for x in data[url]['dispositions']
+            ])
+            page.defendant_types.set([
+                EnforcementActionDefendantType(**x)
+                for x in data[url]['defendant_types']
+            ])
+            page.docket_numbers.set([
+                EnforcementActionDocket(**x)
+                for x in data[url]['docket_numbers']
+            ])
+            page.products.set([
+                EnforcementActionProduct(**x)
+                for x in data[url]['products']
+            ])
+            page.at_risk_groups.set([
+                EnforcementActionAtRisk(**x)
+                for x in data[url]['at_risk_groups']
+            ])
+            page.statutes.set([
+                EnforcementActionStatute(**x)
+                for x in data[url]['statutes']
+            ])
+
+            page.save()
```

Comment:
After some testing, this seems to cause two issues for me:
-  It saves the page even if there are validation errors. So then, after the update, the page can't be saved or published from the Wagtail admin.
- It only saves changes to the current published version of the page. If there is an unpublished draft, it will still have the old metadata, so if someone publishes it after this script, the new published version won't have the updated metadata

I have to head out for the day, but we should look into how to prevent those two issues.

Reply:
Do you have an example of this operating when there are validation errors? Seems like validating the input would be the best way around bullet one. For bullet two, I think we can just handle that case by case. There's only one recent draft that needs to be updated.